File: android/app/src/main/java/com/lockoutprotocol/guardian/data/Prefs.kt
package com.lockoutprotocol.guardian.data

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import java.security.MessageDigest

/**
 * All configuration + secrets live in an encrypted preferences file.
 * Nothing here is readable without the device keystore.
 */
class Prefs private constructor(private val sp: SharedPreferences) {

    // ---- PIN ----
    var pinHash: String?
        get() = sp.getString(K_PIN, null)
        set(v) = sp.edit().putString(K_PIN, v).apply()

    fun setPin(pin: String) { pinHash = sha256(pin) }
    fun checkPin(pin: String): Boolean = pinHash != null && pinHash == sha256(pin)
    fun clearPin() { pinHash = null }
    val pinSet: Boolean get() = pinHash != null

    // ---- Monitored apps (set of package names) ----
    var monitoredPackages: Set<String>
        get() = sp.getStringSet(K_PKGS, emptySet()) ?: emptySet()
        set(v) = sp.edit().putStringSet(K_PKGS, v).apply()

    // ---- Locked apps (set of package names) ----
    var lockedPackages: Set<String>
        get() = sp.getStringSet(K_LOCKED, emptySet()) ?: emptySet()
        set(v) = sp.edit().putStringSet(K_LOCKED, v).apply()

    /**
     * Baseline of user-launchable packages we've already "seen". Used to auto-enroll newly
     * installed apps into [monitoredPackages]: any package present now but not in here is a new
     * install and gets added to monitoring. Removing an app from monitoring requires unlocking
     * Guardian (the whole app is behind the passcode gate), so a fresh install can't dodge it.
     */
    var knownPackages: Set<String>
        get() = sp.getStringSet(K_KNOWN, emptySet()) ?: emptySet()
        set(v) = sp.edit().putStringSet(K_KNOWN, v).apply()

    /**
     * One-time flag: the first run records the current app list as the baseline WITHOUT
     * auto-monitoring everything already installed (those stay user-managed via the picker).
     * Only apps installed after the baseline are auto-enrolled.
     */
    var autoMonitorBaselineDone: Boolean
        get() = sp.getBoolean(K_AUTO_BASE, false)
        set(v) = sp.edit().putBoolean(K_AUTO_BASE, v).apply()

    // ---- Guidelines text fed to the model ----
    var guidelines: String
        get() = sp.getString(K_GUIDE, DEFAULT_GUIDELINES) ?: DEFAULT_GUIDELINES
        set(v) = sp.edit().putString(K_GUIDE, v).apply()

    // ---- Blocking behaviour ----
    var blockMode: String   // "overlay" | "close"
        get() = sp.getString(K_BLOCK, "overlay") ?: "overlay"
        set(v) = sp.edit().putString(K_BLOCK, v).apply()

    /**
     * Fail closed when content can't be seen at all (FLAG_SECURE black frame / capture failure).
     * Default OFF — when on, an always-on monitor can lock you out of your own phone because
     * many normal/system screens read as "unreadable".
     */
    var blockWhenUnreadable: Boolean
        get() = sp.getBoolean(K_BLK_UNREAD, false)
        set(v) = sp.edit().putBoolean(K_BLK_UNREAD, v).apply()

    /** Fail closed when the AI couldn't be reached or its answer couldn't be parsed. Default OFF. */
    var blockWhenUndetermined: Boolean
        get() = sp.getBoolean(K_BLK_UNDET, false)
        set(v) = sp.edit().putBoolean(K_BLK_UNDET, v).apply()

    /**
     * When a monitored app's screen can't be verified (incognito / FLAG_SECURE black frame,
     * capture failure, or the AI being unreachable) — email the accountability contact + log it.
     * This is the SAFE way to catch incognito: it never shows a PIN-locked overlay, so it cannot
     * lock you out of your phone. Default ON.
     */
    var alertOnUnverifiable: Boolean
        get() = sp.getBoolean(K_ALERT_UNVER, true)
        set(v) = sp.edit().putBoolean(K_ALERT_UNVER, v).apply()

    /**
     * Also push the accountability contact for MILD/incidental flags (a swimwear ad, a thumbnail
     * scrolled past) — the ones that produce most of the false positives. Default OFF: those still
     * block and are still logged, they just don't notify. Deliberate flags (tampering, a search, a
     * site for it, explicit content on screen) ALWAYS notify and cannot be silenced from here —
     * see [com.lockoutprotocol.guardian.ai.AlertPolicy].
     */
    var alertOnSuggestive: Boolean
        get() = sp.getBoolean(K_ALERT_SUGG, false)
        set(v) = sp.edit().putBoolean(K_ALERT_SUGG, v).apply()

    /** Additionally send the user Home (escapable nudge) when a monitored app can't be seen. Default OFF. */
    var closeUnverifiable: Boolean
        get() = sp.getBoolean(K_CLOSE_UNVER, false)
        set(v) = sp.edit().putBoolean(K_CLOSE_UNVER, v).apply()

    // ---- Safety covenant (self-binding: make weakening the safeguards visible, not silent) ----
    /** Longest the guidelines have ever been. If they shrink far below this, they were gutted. */
    var guidelinesPeakLen: Int
        get() = sp.getInt(K_COV_PEAK, 0)
        set(v) = sp.edit().putInt(K_COV_PEAK, v).apply()

    /** True once monitoring has been armed (dry-run turned off) at least once. */
    var everArmed: Boolean
        get() = sp.getBoolean(K_COV_ARMED, false)
        set(v) = sp.edit().putBoolean(K_COV_ARMED, v).apply()

    /**
     * Test/dry-run mode: evaluate + log + toast verdicts but NEVER block or lock. Default ON so
     * the app can't lock you out until you've watched the log and deliberately armed it.
     */
    var dryRun: Boolean
        get() = sp.getBoolean(K_DRY_RUN, true)
        set(v) = sp.edit().putBoolean(K_DRY_RUN, v).apply()

    // ---- Push notifications (ntfy — zero-setup alerts) ----
    var ntfyServer: String
        get() = sp.getString(K_NTFY_SRV, "https://ntfy.sh") ?: "https://ntfy.sh"
        set(v) = sp.edit().putString(K_NTFY_SRV, v).apply()

    /** Private alert code (ntfy topic). Auto-generated once; acts as a secret, so keep it private. */
    var ntfyTopic: String
        get() {
            val t = sp.getString(K_NTFY_TOPIC, "") ?: ""
            if (t.isNotBlank()) return t
            val gen = "guardian-" + randomToken(12)
            ntfyTopic = gen
            return gen
        }
        set(v) = sp.edit().putString(K_NTFY_TOPIC, v).apply()

    var pushEnabled: Boolean
        get() = sp.getBoolean(K_PUSH_ON, true)
        set(v) = sp.edit().putBoolean(K_PUSH_ON, v).apply()

    // ---- Ollama Cloud ----
    var ollamaBaseUrl: String
        get() = sp.getString(K_OL_URL, "https://ollama.com") ?: "https://ollama.com"
        set(v) = sp.edit().putString(K_OL_URL, v).apply()
    var ollamaApiKey: String
        get() = sp.getString(K_OL_KEY, "") ?: ""
        set(v) = sp.edit().putString(K_OL_KEY, v).apply()

    /** Backup key. When the active key runs out of quota (or is rejected) the client falls over to
     *  the other one and [activeApiKeySlot] remembers the switch — so the two keys alternate as each
     *  is exhausted, rather than one being tried first forever. */
    var ollamaApiKey2: String
        get() = sp.getString(K_OL_KEY2, "") ?: ""
        set(v) = sp.edit().putString(K_OL_KEY2, v).apply()

    /** Which key is currently preferred: 0 = [ollamaApiKey], 1 = [ollamaApiKey2]. */
    var activeApiKeySlot: Int
        get() = sp.getInt(K_OL_KEY_SLOT, 0).coerceIn(0, 1)
        set(v) = sp.edit().putInt(K_OL_KEY_SLOT, v.coerceIn(0, 1)).apply()

    /**
     * The configured keys, active one first, blanks and duplicates removed. The client walks this
     * list: the first key that isn't out of quota answers the request.
     */
    val apiKeys: List<String>
        get() {
            val slots = listOf(ollamaApiKey.trim(), ollamaApiKey2.trim())
            val ordered = if (activeApiKeySlot == 1) slots.asReversed() else slots
            return ordered.filter { it.isNotBlank() }.distinct()
        }

    /** Remember that [key] is the one currently working, so the next request starts there. */
    fun promoteApiKey(key: String) {
        val slot = when (key) {
            ollamaApiKey.trim() -> 0
            ollamaApiKey2.trim() -> 1
            else -> return
        }
        if (activeApiKeySlot != slot) activeApiKeySlot = slot
    }

    var ollamaModel: String
        get() = sp.getString(K_OL_MODEL, DEFAULT_MODEL) ?: DEFAULT_MODEL
        set(v) = sp.edit().putString(K_OL_MODEL, v).apply()

    // ---- Heartbeat (optional tamper server) ----
    var heartbeatUrl: String get() = sp.getString(K_HB_URL, "") ?: ""
        set(v) = sp.edit().putString(K_HB_URL, v).apply()

    // ---- Runtime state ----
    var monitoringEnabled: Boolean get() = sp.getBoolean(K_ENABLED, false)
        set(v) = sp.edit().putBoolean(K_ENABLED, v).apply()
    var lastViolationAt: Long get() = sp.getLong(K_LASTVIO, 0)
        set(v) = sp.edit().putLong(K_LASTVIO, v).apply()

    // ---- Tamper resistance ----
    /**
     * Last-known state of Guardian's accessibility access. The heartbeat + service teardown compare
     * against this so a silent "accessibility turned off" (the way monitoring was defeated before)
     * is detected and reported to the accountability contact instead of failing quietly.
     */
    var accessibilityGranted: Boolean get() = sp.getBoolean(K_A11Y_GRANTED, false)
        set(v) = sp.edit().putBoolean(K_A11Y_GRANTED, v).apply()

    /** Show the pledge overlay whenever the system Settings app is opened. Default ON. */
    var pledgeOnSettings: Boolean get() = sp.getBoolean(K_PLEDGE, true)
        set(v) = sp.edit().putBoolean(K_PLEDGE, v).apply()

    /**
     * Self-guard: while accessibility is still on, detect the specific Settings screens used to
     * disable Guardian (its accessibility toggle, its App-info / force-stop / uninstall page) and
     * bounce out + alert. Only targets Guardian's own screens, so the rest of Settings stays usable.
     * Default ON.
     */
    var guardSelf: Boolean get() = sp.getBoolean(K_GUARD_SELF, true)
        set(v) = sp.edit().putBoolean(K_GUARD_SELF, v).apply()

    companion object {
        private const val FILE = "guardian_secure_prefs"
        private const val K_PIN = "pin_hash"
        private const val K_PKGS = "monitored_pkgs"
        private const val K_LOCKED = "locked_pkgs"
        private const val K_KNOWN = "known_pkgs"
        private const val K_AUTO_BASE = "auto_monitor_baseline_done"
        private const val K_GUIDE = "guidelines"
        private const val K_BLOCK = "block_mode"
        private const val K_BLK_UNREAD = "block_unreadable"
        private const val K_BLK_UNDET = "block_undetermined"
        private const val K_ALERT_UNVER = "alert_unverifiable"
        private const val K_ALERT_SUGG = "alert_suggestive"
        private const val K_CLOSE_UNVER = "close_unverifiable"
        private const val K_COV_PEAK = "covenant_guidelines_peak_len"
        private const val K_COV_ARMED = "covenant_ever_armed"
        private const val K_DRY_RUN = "dry_run"
        private const val K_NTFY_SRV = "ntfy_server"
        private const val K_NTFY_TOPIC = "ntfy_topic"
        private const val K_PUSH_ON = "push_enabled"

        private fun randomToken(n: Int): String {
            val chars = "abcdefghijklmnopqrstuvwxyz0123456789"
            return (1..n).map { chars.random() }.joinToString("")
        }
        private const val K_OL_URL = "ollama_url"
        private const val K_OL_KEY = "ollama_key"
        private const val K_OL_KEY2 = "ollama_key2"
        private const val K_OL_KEY_SLOT = "ollama_key_slot"
        private const val K_OL_MODEL = "ollama_model"
        // Vision-capable Gemma model hosted on Ollama Cloud.
        const val DEFAULT_MODEL = "gemma4:31b-cloud"

        private const val K_HB_URL = "hb_url"
        private const val K_ENABLED = "enabled"
        private const val K_LASTVIO = "last_violation"
        private const val K_A11Y_GRANTED = "a11y_granted"
        private const val K_PLEDGE = "pledge_on_settings"
        private const val K_GUARD_SELF = "guard_self"

        val DEFAULT_GUIDELINES = """
            Prohibited content — be STRICT. Flag anything sexual or sexually suggestive:
            - Pornography and sexually explicit imagery or acts.
            - Real nudity OR partial nudity.
            - Women or men in underwear, lingerie, bikinis, or swimwear.
            - Exposed or emphasised skin in sensual areas: cleavage, breasts, midriff, buttocks,
              crotch, or bare thighs; see-through, wet, or tight clothing that sexualises the body.
            - Sexualised or provocative poses, or close-ups of intimate areas.
            - Any image a person could reasonably be aroused by or masturbate to.

            Also flag INTENT: if a search query, typed text, URL, or title shows the user is trying
            to find or view such imagery — including searching Google, Google Images, an image site,
            or YouTube for sexual / nude / "hot" / "sexy" / bikini / lingerie content — flag it and
            quote the text.

            Only clearly innocent, non-sexual screens are allowed. When an image is even borderline
            sexual, treat it as a violation rather than giving the benefit of the doubt.
        """.trimIndent()

        @Volatile private var instance: Prefs? = null

        fun get(context: Context): Prefs = instance ?: synchronized(this) {
            instance ?: build(context.applicationContext).also { instance = it }
        }

        private fun build(ctx: Context): Prefs {
            val key = MasterKey.Builder(ctx)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            val sp = EncryptedSharedPreferences.create(
                ctx, FILE, key,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
            return Prefs(sp)
        }

        fun sha256(s: String): String =
            MessageDigest.getInstance("SHA-256").digest(s.toByteArray())
                .joinToString("") { "%02x".format(it) }
    }
}

File: macos/Guardian/Models/Prefs.swift
import CryptoKit
import Foundation

/// All configuration lives in UserDefaults; secrets (Ollama key, ntfy topic, PIN hash) live in the
/// macOS Keychain. Mirrors the Android `Prefs` (which used EncryptedSharedPreferences).
///
/// `ObservableObject` so SwiftUI views update live.
@MainActor
final class Prefs: ObservableObject {

    static let shared = Prefs()
    private let d = UserDefaults.standard

    // MARK: - Monitored apps (set of bundle identifiers)
    @Published var monitoredApps: Set<String> { didSet { d.set(Array(monitoredApps), forKey: K.monitored) } }

    // MARK: - Guidelines fed to the model
    @Published var guidelines: String { didSet { d.set(guidelines, forKey: K.guidelines) } }

    // MARK: - Blocking behaviour
    /// "hide" (send the app to the background — escapable) or "quit" (terminate the offending app).
    @Published var blockMode: String { didSet { d.set(blockMode, forKey: K.blockMode) } }

    /// Test/dry-run mode: evaluate + log verdicts but NEVER hide/quit anything. Default ON.
    @Published var dryRun: Bool { didSet { d.set(dryRun, forKey: K.dryRun) } }

    /// Push the accountability contact + log when a monitored app's screen can't be verified
    /// (capture failure, blank frame, or the AI being unreachable). Never quits. Default ON.
    @Published var alertOnUnverifiable: Bool { didSet { d.set(alertOnUnverifiable, forKey: K.alertUnver) } }

    /// Additionally hide the app when its screen can't be verified. Default OFF.
    @Published var closeUnverifiable: Bool { didSet { d.set(closeUnverifiable, forKey: K.closeUnver) } }

    // MARK: - Push notifications (ntfy — zero-setup alerts)
    @Published var ntfyServer: String { didSet { d.set(ntfyServer, forKey: K.ntfyServer) } }
    @Published var pushEnabled: Bool { didSet { d.set(pushEnabled, forKey: K.pushEnabled) } }

    /// Private alert code (ntfy topic). Auto-generated once; acts as a secret, so keep it private.
    var ntfyTopic: String {
        get {
            if let t = Keychain.get(K.ntfyTopic), !t.isEmpty { return t }
            let gen = "guardian-" + Self.randomToken(12)
            Keychain.set(K.ntfyTopic, gen)
            return gen
        }
        set { Keychain.set(K.ntfyTopic, newValue) }
    }

    // MARK: - Ollama Cloud
    @Published var ollamaBaseUrl: String { didSet { d.set(ollamaBaseUrl, forKey: K.ollamaUrl) } }
    @Published var ollamaModel: String { didSet { d.set(ollamaModel, forKey: K.ollamaModel) } }
    /// Secret — stored in Keychain, not UserDefaults.
    var ollamaApiKey: String {
        get { Keychain.get(K.ollamaKey) ?? "" }
        set { Keychain.set(K.ollamaKey, newValue); objectWillChange.send() }
    }

    /// Backup key. When the active key runs out of quota (or is rejected) `OllamaClient` falls over
    /// to the other one and `activeApiKeySlot` remembers the switch — so the two keys alternate as
    /// each is exhausted, rather than one being tried first forever.
    var ollamaApiKey2: String {
        get { Keychain.get(K.ollamaKey2) ?? "" }
        set { Keychain.set(K.ollamaKey2, newValue); objectWillChange.send() }
    }

    /// Which key is currently preferred: 0 = `ollamaApiKey`, 1 = `ollamaApiKey2`.
    var activeApiKeySlot: Int {
        get { min(max(d.integer(forKey: K.ollamaKeySlot), 0), 1) }
        set { d.set(min(max(newValue, 0), 1), forKey: K.ollamaKeySlot) }
    }

    /// The configured keys, active one first, blanks and duplicates removed. The client walks this
    /// list: the first key that isn't out of quota answers the request.
    var apiKeys: [String] {
        let slots = [ollamaApiKey.trimmed(), ollamaApiKey2.trimmed()]
        let ordered = activeApiKeySlot == 1 ? slots.reversed().map { $0 } : slots
        var seen = Set<String>()
        return ordered.filter { !$0.isEmpty && seen.insert($0).inserted }
    }

    /// Remember that `key` is the one currently working, so the next request starts there.
    func promoteApiKey(_ key: String) {
        let slot: Int
        switch key {
        case ollamaApiKey.trimmed(): slot = 0
        case ollamaApiKey2.trimmed(): slot = 1
        default: return
        }
        if activeApiKeySlot != slot { activeApiKeySlot = slot }
    }

    // MARK: - Access / override PIN (SHA-256 hash kept in Keychain)
    /// Gates opening Guardian AND dismissing a violation block screen. Optional — when unset, the
    /// app opens freely and a block screen can be dismissed with a plain button.
    var pinSet: Bool { Keychain.get(K.pin) != nil }
    func setPin(_ pin: String) { Keychain.set(K.pin, Self.sha256(pin)); objectWillChange.send() }
    func clearPin() { Keychain.delete(K.pin); objectWillChange.send() }
    func checkPin(_ pin: String) -> Bool {
        guard let h = Keychain.get(K.pin) else { return false }
        return h == Self.sha256(pin)
    }

    static func sha256(_ s: String) -> String {
        SHA256.hash(data: Data(s.utf8)).map { String(format: "%02x", $0) }.joined()
    }

    // MARK: - Tamper resistance
    /// Show the pledge overlay whenever System Settings is brought frontmost.
    @Published var pledgeOnSettings: Bool { didSet { d.set(pledgeOnSettings, forKey: K.pledge) } }

    /// Last-known state of the Screen Recording (TCC) grant. The tamper guard compares against this
    /// so revoking it — the way monitoring is defeated on macOS — is detected and reported instead
    /// of failing silently.
    @Published var screenRecordingGranted: Bool { didSet { d.set(screenRecordingGranted, forKey: K.scrGranted) } }

    /// Keep Guardian running at login (persistence, so quitting/killing it isn't a lasting bypass).
    @Published var relaunchAtLogin: Bool { didSet { d.set(relaunchAtLogin, forKey: K.relaunch) } }

    /// Stronger persistence: a KeepAlive LaunchAgent that relaunches Guardian within ~10s of ANY
    /// exit (Quit / force-quit / crash), not just at login. Supersedes [relaunchAtLogin] when on.
    /// Defaults ON — quitting Guardian is a bypass, so the app resists it out of the box; turn it
    /// off from Settings ▸ Tamper resistance (behind the passcode) before quitting legitimately.
    @Published var keepAlive: Bool { didSet { d.set(keepAlive, forKey: K.keepAlive) } }

    // MARK: - Safety covenant (self-binding: make weakening the safeguards visible, not silent)
    /// Longest the guidelines have ever been; a big drop below this means they were gutted.
    @Published var guidelinesPeakLen: Int { didSet { d.set(guidelinesPeakLen, forKey: K.covPeak) } }
    /// True once monitoring has been armed (dry-run turned off) at least once.
    @Published var everArmed: Bool { didSet { d.set(everArmed, forKey: K.covArmed) } }

    // MARK: - Runtime state
    @Published var monitoringEnabled: Bool { didSet { d.set(monitoringEnabled, forKey: K.enabled) } }
    var lastViolationAt: Date? {
        get { (d.object(forKey: K.lastVio) as? Double).map { Date(timeIntervalSince1970: $0) } }
        set { d.set(newValue?.timeIntervalSince1970, forKey: K.lastVio) }
    }

    // MARK: - Init (hydrate from storage with defaults)

    private init() {
        let dd = UserDefaults.standard
        monitoredApps = Set(dd.stringArray(forKey: K.monitored) ?? [])
        guidelines = dd.string(forKey: K.guidelines) ?? Self.defaultGuidelines
        blockMode = dd.string(forKey: K.blockMode) ?? "hide"
        dryRun = dd.object(forKey: K.dryRun) as? Bool ?? true
        alertOnUnverifiable = dd.object(forKey: K.alertUnver) as? Bool ?? true
        closeUnverifiable = dd.object(forKey: K.closeUnver) as? Bool ?? false
        ntfyServer = dd.string(forKey: K.ntfyServer) ?? "https://ntfy.sh"
        pushEnabled = dd.object(forKey: K.pushEnabled) as? Bool ?? true
        ollamaBaseUrl = dd.string(forKey: K.ollamaUrl) ?? "https://ollama.com"
        ollamaModel = dd.string(forKey: K.ollamaModel) ?? Self.defaultModel
        monitoringEnabled = dd.object(forKey: K.enabled) as? Bool ?? false
        pledgeOnSettings = dd.object(forKey: K.pledge) as? Bool ?? true
        screenRecordingGranted = dd.object(forKey: K.scrGranted) as? Bool ?? false
        relaunchAtLogin = dd.object(forKey: K.relaunch) as? Bool ?? true
        keepAlive = dd.object(forKey: K.keepAlive) as? Bool ?? true
        guidelinesPeakLen = dd.integer(forKey: K.covPeak)
        everArmed = dd.object(forKey: K.covArmed) as? Bool ?? false

        // Remember once we've ever been armed, so the covenant can notice a later revert to TEST mode.
        if !dryRun { everArmed = true }
    }

    // MARK: - Keys

    private enum K {
        static let monitored = "monitored_apps"
        static let guidelines = "guidelines"
        static let blockMode = "block_mode"
        static let dryRun = "dry_run"
        static let alertUnver = "alert_unverifiable"
        static let closeUnver = "close_unverifiable"
        static let ntfyServer = "ntfy_server"
        static let ntfyTopic = "ntfy_topic"          // keychain
        static let pushEnabled = "push_enabled"
        static let ollamaUrl = "ollama_url"
        static let ollamaModel = "ollama_model"
        static let ollamaKey = "ollama_key"           // keychain
        static let ollamaKey2 = "ollama_key2"         // keychain
        static let ollamaKeySlot = "ollama_key_slot"
        static let pin = "pin_hash"                    // keychain
        static let enabled = "monitoring_enabled"
        static let lastVio = "last_violation_at"
        static let pledge = "pledge_on_settings"
        static let scrGranted = "screen_recording_granted"
        static let relaunch = "relaunch_at_login"
        static let keepAlive = "keep_alive_agent"
        static let covPeak = "covenant_guidelines_peak_len"
        static let covArmed = "covenant_ever_armed"
    }

    // Vision-capable Gemma 4 hosted on Ollama Cloud (multimodal, 256K context).
    static let defaultModel = "gemma4:31b-cloud"

    static let defaultGuidelines = """
        Prohibited content — anything sexual or sexually suggestive:
        - Pornography and sexually explicit imagery or acts.
        - Real nudity OR partial nudity.
        - Women or men in underwear, lingerie, bikinis, or swimwear.
        - Exposed or emphasised skin in sensual areas: cleavage, breasts, midriff, buttocks, crotch,
          or bare thighs; see-through, wet, or tight clothing that sexualises the body.
        - Sexualised or provocative poses, or close-ups of intimate areas.
        - Any image a person could reasonably be aroused by or masturbate to.

        Also flag INTENT: if a search query, typed text, URL, or title shows the user is trying to
        find or view such imagery — including searching Google, Google Images, an image site, or
        YouTube for sexual / nude / "hot" / "sexy" / bikini / lingerie content — flag it and quote
        the text.

        Ordinary clothed people, fashion, art, and educational/medical content are not violations on
        their own — but when an image is borderline, it is treated as a violation rather than given
        the benefit of the doubt.
        """

    private static func randomToken(_ n: Int) -> String {
        let chars = Array("abcdefghijklmnopqrstuvwxyz0123456789")
        return String((0..<n).map { _ in chars.randomElement()! })
    }
}

private extension String {
    func trimmed() -> String { trimmingCharacters(in: .whitespacesAndNewlines) }
}
